[Mobile Payments] Disconnect built-in reader when Manage card readers is tapped

joshheald opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: #8664
<!-- Id number of the GitHub issue this PR addresses. -->

## Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->

In sTAP Away, we add support for the built-in card reader, allowing merchants to take card payments using Tap to Pay on iPhone.

We can be connected to either the built-in reader, **or** a Bluetooth reader, but not both simultaneously.

There’s nothing to manage or view about the built in reader, and from a user’s perspective there isn’t a reader at all, it’s just part of their phone.

Since we can’t connect to the Bluetooth reader until the built in reader is disconnected, we just do that automatically for the user here. This saves us updating the display of the connected reader to be suitable for the built in reader too.

We can revisit this in future, but for M1 it makes sense.

## Testing instructions
<!-- Step by step testing instructions. When necessary break out individual scenarios that need testing, consider including a checklist for the reviewer to go through. -->

### Using a US-based WCPay or Stripe store:

1. Navigate to `Menu > Settings > Experimental features`
2. Turn on `Tap to Pay on iPhone`
3. Navigate to `Menu > Payments > Collect payment`
4. Go through the payment flow, and select `Card` on the payment method screen
5. When asked for a reader type, tap `Tap to Pay on iPhone` and connect to the reader
6. Take a payment. 
7. When it's finished, repeat the above three steps to take another payment: observe that you don't need to connect to the reader and are not asked for a reader type.
8. Go to `Menu > Payments > Manage Card Reader`
9. Observe that there is no reader shown as connected
10. Repeat the above steps to take a payment: observe that you are asked to choose a reader type again, and can connect to the reader as the first time.

In general –
1. Observe that TTPoI connections persist as long as the app is in the foreground, unless you tap `Manage Card Reader`
2. Observe that Bluetooth connections persist until you tap `Disconnect` in the `Manage Card Reader` screen.



## Screenshots
<!-- Include before and after images or gifs when appropriate. -->

https://user-images.githubusercontent.com/2472348/212901722-3e60defe-336e-435b-bcc7-9c68b2b13f05.mp4



---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

joshheald committed:
8664 Disconnect built-in reader when Manage tapped

There’s nothing to manage or view about the built in reader, and from a user’s perspective there isn’t a reader at all, it’s just part of their phone.

Since we can’t connect to the Bluetooth reader until the built in reader is disconnected, we just do that automatically for the user here.

This saves us updating the display of the connected reader to be suitable for the built in reader too.

We can revisit this in future, but for M1 it makes sense.

toupper commented on WooCommerce/Classes/ViewRelated/Dashboard/Settings/CardReadersV2/CardReaderSettingsConnectedViewModel.swift:
```diff
@@ -146,12 +146,21 @@ final class CardReaderSettingsConnectedViewModel: CardReaderSettingsPresentedVie
     }
 
     private func updateProperties() {
+        if let connectedReader = connectedReaders.first,
```
I am wondering, is this the best place to disconnect the reader? After all, the purpose of this function is to update the properties with the first connected reader data; disconnecting the reader seems like an unexpected side effect here.

joshheald replied:
Fair point: I could put it directly in the completion block for `observeConnectedReaders` in `beginObservation`. It won't make any difference to the functionality or way it works though: `updateProperties` is only called from that same block.

joshheald replied:
Moved in 0a92d0f

toupper commented on WooCommerce/Classes/ViewRelated/Dashboard/Settings/CardReadersV2/CardReaderSettingsConnectedViewModel.swift:
```diff
@@ -249,6 +258,8 @@ final class CardReaderSettingsConnectedViewModel: CardReaderSettingsPresentedVie
             newShouldShow = .isUnknown
         } else if connectedReaders.isEmpty {
             newShouldShow = .isFalse
+        } else if connectedReaders.first(where: { $0.readerType == .appleBuiltIn }) != nil {
```
Technically this means that if any of the connected readers is the Apple Built-in one we don't show the view right? What happens if we are connected to a Bluetooth reader apart from of that?

joshheald replied:
Correct, however we don't actually support being connected to more than one reader, and [neither do Stripe](https://stripe.dev/stripe-terminal-ios/docs/Classes/SCPTerminal.html#/c:objc(cs)SCPTerminal(py)connectedReader).

So we can't be connected to more than one at a time, and this class doesn't otherwise handle that situation anyway (it relies on `.first` to determine the connected reader.)

joshheald replied:
The proper fix for this inconsistency would be to change `connectedReaders` to a single object, not an array, but that is way out of scope here 😊

toupper replied:
> The proper fix for this inconsistency would be to change connectedReaders to a single object, not an array, but that is way out of scope here 😊

That's what I thought, can we create an issue to change that?

toupper commented on WooCommerce/WooCommerceTests/Mocks/MockCardPresentPaymentsStoresManager.swift:
```diff
@@ -19,6 +19,8 @@ final class MockCardPresentPaymentsStoresManager: DefaultStoresManager {
     private var softwareUpdateSubject: CurrentValueSubject<CardReaderSoftwareUpdateState, Never> = .init(.none)
     private var paymentExtension: CardPresentPaymentGatewayExtension
 
+    var recievedActions: [CardPresentPaymentAction] = []
```
Typo here, `recievedActions` -> `receivedActions`

joshheald replied:
Will fix

joshheald replied:
Fixed in 791b53b

toupper commented:
Thanks for the PR @joshheald! I have a question regarding the implementation, also link with my first comment.

If I follow the flow correctly, every time the list of connected readers changes we are notified by observing the `observeConnectedReaders` action completion block. Once it changes, we call to update the properties, where we disconnect the first reader if it's the Apple Built-in one.

Is it like that? In that case, what happens if the list changes because the Apple Built-in reader is just connected? From this perspective, we can end up having the opposite effect we expected: We wanted to connect the Apple Built-in reader, but it ends up being disconnected.

joshheald commented:
@toupper Thanks for the review
> Is it like that? In that case, what happens if the list changes because the Apple Built-in reader is just connected? From this perspective, we can end up having the opposite effect we expected: We wanted to connect the Apple Built-in reader, but it ends up being disconnected.

This is not a problem: the code in question is specific to the `Manage card reader` screen, and we do not allow merchants to connect to the built-in reader from that screen. The `Connect card reader` button on there will _only_ trigger a Bluetooth scan, not a built-in reader connection.

This is because the merchant is unlikely to think of the built-in reader as a "Card reader" they need to manage, it's just a part of their phone.

joshheald committed:
8664 Fix typo in mock

joshheald committed:
Move built-in reader disconnect so it’s not a side effect

toupper commented:
> This is not a problem: the code in question is specific to the Manage card reader screen, and we do not allow merchants to connect to the built-in reader from that screen. The Connect card reader button on there will only trigger a Bluetooth scan, not a built-in reader connection.
> 
> This is because the merchant is unlikely to think of the built-in reader as a "Card reader" they need to manage, it's just a part of their phone.

Thanks for your answer @joshheald ! That makes sense now. I have a few suggestions to make this more clear:

- Rename `CardReaderSettingsConnectedViewModel` to `BluetoothCardReaderSettingsConnectedViewModel`. That way we know that that screen is only for Bluetooth readers.
- Move the code 
```
if readers.first(where: { $0.readerType == .appleBuiltIn }) != nil {
                self.disconnect()
            }
```
to its own function, with naming `disconnectAppleBuiltInReader`, so we see what's being done here.
- Add a comment in that function to describe why we need to do that

toupper commented:
@joshheald The code looks good and it tests well 🎉 I added a few code suggestions, and I have one small finding when testing:

> - Repeat the above steps to take a payment: observe that you are asked to choose a reader type again, and can connect to the reader as the first time.

When testing this step, not only are we asked what reader we want to connect, but we also show a loading screen with "_Connecting to your account_" message. If the reader is connected (step 7) we don't need to connect to our account. Why is being disconnected from a reader causes having to connect to the account?

joshheald commented:
Thanks for the re-review @toupper

> When testing this step, not only are we asked what reader we want to connect, but we also show a loading screen with "_Connecting to your account_" message. If the reader is connected (step 7) we don't need to connect to our account. Why is being disconnected from a reader causes having to connect to the account?

This happens in the onboarding check, it's pre-existing. We always check the onboarding status before any reader connection, so that we can be sure the user's account is in the correct state to connect/take a payment, but we don't do that if they're already connected to a reader as we can presume that they've had the check done recently.

joshheald committed:
8664 ReaderSettingsConnectedVM is Bluetooth specific

The CardReaderSettingsConnectedViewModel is now specific to Bluetooth readers, as we do not expect users to view the built-in card reader as something they need to connect to and manage.

This commit neatens that up and explains the logic.

joshheald commented:
@toupper Bluetooth-specific updates in e3f9fcc, thanks for the suggestions
